src/rust/ide/lib/enso-protocol/src/language_server/types.rs

```diff
@@ -462,4 +475,130 @@ pub enum RegisterOptions {
     Path {path:Path},
     #[serde(rename_all = "camelCase")]
     ExecutionContextId {context_id:ContextId},
+    None {},
+}
+
+
+// ===========================
+// === Suggestion Database ===
+// ===========================
+
+/// The identifier of SuggestionEntry in SuggestionDatabase.
+pub type SuggestionEntryId = usize;
+
+/// The version of Suggestion Database.
+pub type SuggestionsDatabaseVersion = usize;
+
+/// The argument of an atom, method or function suggestion.
+#[derive(Hash, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SuggestionEntryArgument {
+    /// The argument name.
+    pub name: String,
+    /// The arguement type. String 'Any' is used to specify generic types.
+    #[serde(rename = "type")] // To avoid collision with the `type` keyword.
+    pub arg_type: String,
+    /// Indicates whether the argument is lazy.
+    pub is_suspended: bool,
+    /// Optional default value.
+    pub default_value: Option<String>,
+}
+
+/// The definition scope. The start and end are chars indices.
+#[derive(Hash,Debug,Copy,Clone,PartialEq,Eq,Serialize,Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(missing_docs)]
+pub struct SuggestionEntryScope {
+    pub start : usize,
+    pub end   : usize,
+}
+
+/// A type of suggestion entry.
+#[derive(Hash,Debug,Copy,Clone,PartialEq,Eq,Serialize,Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(missing_docs)]
+pub enum SuggestionEntryType {Atom,Method,Function,Local}
+
+/// A Suggestion Entry.
+#[derive(Hash, Debug, Clone, PartialEq, Eq,Serialize,Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(missing_docs)]
+pub enum SuggestionEntry {
+    SuggestionEntryAtom {
+        name          : String,
+        module        : String,
+        arguments     : Vec<SuggestionEntryArgument>,
+        return_type   : String,
+        documentation : Option<String>,
+    },
+    SuggestionEntryMethod {
+        name          : String,
+        module        : String,
+        arguments     : Vec<SuggestionEntryArgument>,
+        self_type     : String,
+        return_type   : String,
+        documentation : Option<String>,
+    },
+    SuggestionEntryFunction {
+        name        : String,
+        module      : String,
+        arguments   : Vec<SuggestionEntryArgument>,
+        return_type : String,
+        scope       : SuggestionEntryScope,
+    },
+    SuggestionEntryLocal {
+        name        : String,
+        module      : String,
+        return_type : String,
+        scope       : SuggestionEntryScope,
+    },
+}
+
+impl SuggestionEntry {
+    /// Get name of the suggested entity.
+    pub fn name(&self) -> &String {
+        match self {
+            Self::SuggestionEntryAtom     {name,..} => name,
+            Self::SuggestionEntryFunction {name,..} => name,
+            Self::SuggestionEntryLocal    {name,..} => name,
+            Self::SuggestionEntryMethod   {name,..} => name,
+        }
+    }
+}
```

Comment:
Why this is not just a nested structure with name field refactored as common field?

Reply:
For the rule "struct should reflect the API". But I will see how we will use it, if it became troublesome, perhaps I create another structure, or refactor this one.